ebpf/bpf/pid.h

```diff
@@ -1,47 +1,42 @@
-
-#if !defined(PYROSCOPE_PID)
-#define PYROSCOPE_PID
-
-// this should not be used in production, and always be disabled
-// but is useful for running in a privileged context outside host pid namespace, for example wsl2
-//#define PYROSCOPE_PID_NAMESPACED
-
-#if defined(PYROSCOPE_PID_NAMESPACED)
-
-#include "bpf_core_read.h"
-// https://github.com/grafana/beyla/blob/6366275ce2d2c9bdefd47975b389fbcf39cbbea8/bpf/pid.h#L13
-// Good resource on this: https://mozillazg.com/2022/05/ebpf-libbpfgo-get-process-info-en.html
-// Using bpf_get_ns_current_pid_tgid is too restrictive for us
-//static __always_inline void ns_pid_ppid(struct task_struct *task, u32 *pid , int *ppid, u32 *pid_ns_id) {
-static __always_inline void current_pid(u32 *pid) {
-    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
-    if (task == 0) {
-        return;
-    }
-    struct upid upid;
-
-    unsigned int level = BPF_CORE_READ(task, nsproxy, pid_ns_for_children, level);
-    struct pid *ns_pid = (struct pid *)BPF_CORE_READ(task, group_leader, thread_pid);
-    bpf_probe_read_kernel(&upid, sizeof(upid), &ns_pid->numbers[level]);
-
-    *pid = (u32)upid.nr;
-//    unsigned int p_level = BPF_CORE_READ(task, real_parent, nsproxy, pid_ns_for_children, level);
-//
-//    struct pid *ns_ppid = (struct pid *)BPF_CORE_READ(task, real_parent, group_leader, thread_pid);
-//    bpf_probe_read_kernel(&upid, sizeof(upid), &ns_ppid->numbers[p_level]);
-//    *ppid = upid.nr;
-//
-//    struct ns_common ns = BPF_CORE_READ(task, nsproxy, pid_ns_for_children, ns);
-//    *pid_ns_id = ns.inum;
-}
-
-#else // PYROSCOPE_PID_NAMESPACED
-
-static __always_inline void current_pid(u32 *pid) {
-  u64 pid_tgid = bpf_get_current_pid_tgid();
-  *pid = (u32)(pid_tgid >> 32);
-}
-#endif // PYROSCOPE_PID_NAMESPACED
-
-
-#endif // PYROSCOPE_PID
\ No newline at end of file
+#if !defined(PYROSCOPE_PID)
+#define PYROSCOPE_PID
+
+#include "bpf_core_read.h"
+#include "bpf_helpers.h"
+#include "vmlinux.h"
+
+#define PID_NESTED_NAMESPACES_MAX 4
+
+static __always_inline void current_pid(uint64_t ns_pid_ino, uint32_t *pid) {
+    struct upid upid;
+    unsigned int inum;
+
+    // fallback to host pid, if no inode provided
+    if (ns_pid_ino == 0) {
+        uint64_t pid_tgid = bpf_get_current_pid_tgid();
+        *pid = (u32)(pid_tgid >> 32);
+        return;
+    }
+
+    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
+
+    // retrieve level nested namespaces
+    unsigned int level = BPF_CORE_READ(task, nsproxy, pid_ns_for_children, level);
+
+    // match the level with pid ns inode
+#pragma unroll
+    for (int i = 0; i < PID_NESTED_NAMESPACES_MAX; i++) {
+        if ((level - i) < 0) {
+            break;
+        }
+        upid = BPF_CORE_READ(task, thread_pid, numbers[level - i]);
+        inum = BPF_CORE_READ(upid.ns, ns.inum);
+
+        if (inum == ns_pid_ino) {
+            *pid = upid.nr;
```

Comment:
Should we add BPF_CORE_READ here as well? in case struct upid changes

Reply:
I am not quite sure, if you can do that. I've only used `BPF_CORE_READ` for fetching pointers.

Reply:
not 100 % sure but maybe like this
```C
inum = BPF_CORE_READ(task, thread_pid, numbers[level - i].ns, ns.inum);
if (inum == ns_pid_inode) {
    *pid = BPF_CORE_READ(task, thread_pid, numbers[level - i].nr);
```

Reply:
I also noticed that .nr field is a PID in kernel terms aka TID in userspace terms 
And ti should be TGID in kernel terms and PID in userspace terms
So it seems incorrect

Reply:
On `upid.nr` I do think it is a tgid (rather than a kernel thread pid), look at those links (this is where I got it from). It also seems to match the correct processes, when I run the playground (and containers):

- https://mozillazg.com/2022/05/ebpf-libbpfgo-get-process-info-en.html
- https://github.com/grafana/beyla/blob/d225274a9d00e83313fde64e433e9a91179ba16e/bpf/pid_types.h#L19C30-L37

Reply:
hmm im confused

both links seem to touch group_leader and in this PR we  dont, maybe we should just do ?

```C
unsigned int level = BPF_CORE_READ(task, group_leader, nsproxy, pid_ns_for_children, level);
```

Reply:
Yes you completely right, 23bb5b1